### ISSUE TITLE
coupled() and coupledScalar() can be mixed up without errors

### ISSUE THREAD

**Issue by permcody** [idaholab/moose#10394]
## Rationale

Rui Hu at Argonne was getting several segfaults in his application and mentioned to me that he fixed the errors by finding out he was using `coupled()` instead of `coupledScalar()` in a few spots. These functions should definitely be aborting when they are used correctly because otherwise they'll be returning references to memory locations that will either not be initialized or will be improperly sized.

## Description

We do not have examples of this currently but it should be straightforward to create an input file with scalar and spatial variables that then retrieve references to them using the wrong function. Proper test cases will need to be created and added to the regression test suite.

## Impact

Framework robustness

FYI - @andrsd

**Comment by andrsd** [comment]
If this is only in regard to `coupled` and `coupledScalar`, they return variable number, not memory. 

I am wondering why it did not error out, because in `coupled` we call `getVar` and it looks into `_coupled_vars` for the variable name. This container has only `MooseVariable`s, so it should fail looking up a scalar variable and we should hit the moose error saying "Trying to get a non-existent variable".

If this is also in regard to `coupledValue` et al. The same `getVar` call is made there, so again the code should give an error...

**Comment by andrsd** [comment]
So this input file si coupling a scalar variable where field variable is expected:
```
[Mesh]
  type = GeneratedMesh
  dim = 2
  nx = 2
  ny = 2
[]

[Variables]
  [./u]
  [../]
  [./a]
    family = SCALAR
    order = FIRST
  [../]
[]

[Kernels]
  [./diff]
    type = Diffusion
    variable = u
  [../]

  [./coupled]
    type = CoupledForce
    variable = u
    v = a
  [../]
[]

[ScalarKernels]
  [./alpha]
    type = AlphaCED
    variable = a
    value = 1
  [../]
[]

[BCs]
  [./all]
    type = DirichletBC
    boundary = 'left right top bottom'
    variable = u
    value = 0
  [../]
[]

[Executioner]
  type = Steady
[]
```
and MOOSE errors out with:
```
Attempted to retrieve default value for coupled variable 'v' when none was provided. 

There are three reasons why this may have occurred:
 1. The other version of params.addCoupledVar() should be used in order to provide a default value. 
 2. This should have been a required coupled variable added with params.addRequiredCoupledVar() 
 3. The call to get the coupled value should have been properly guarded with isCoupled()
```
That's not really the reason why the input file is wrong, so I think we can improve the logic and give back a better error message. But no segfault...

**Comment by permcody** [comment]
@andrsd - Closing this since it lacks a clear definition of a repeatable problem and you have addressed/improved error checking in this general area of the setup reasonable well. We can always re-open or create a new ticket if we encounter a specific new problem.

Closed #10398
